File: DSLib/Client/BinaryTree/BinaryTree.cpp
#include "BST.h"

#include <iostream>
#include <vector>

int main(int argc, char* argv[])
{
	ds::bin_tree::bst<int> bst;
	std::vector<int> data{ 3, 1, 2, 4, 5, 0, 6 };
	for (auto item : data)
	{
		bst.insert(item);
	}

	std::vector<int> inorder;

	// Inorder - specifying a custom visit action
	bst.inorder([&inorder](std::shared_ptr<ds::bin_tree::node<int>> n) {inorder.push_back(n->data); });
	std::cout << "Inorder:" << std::endl;
	for (const auto &item : inorder)
	{
		std::cout << item << std::endl;
	}

	// Preorder
	std::cout << "\nPreorder:" << std::endl;
	bst.preorder();

	// Postorder
	std::cout << "\nPostorder:" << std::endl;
	bst.postorder();

	// Height of the tree
	std::cout << "\nHeight: " << bst.height() << std::endl;
	return 0;
}

File: DSLib/Client/BinaryTree/BinaryTree.java
package Client.BinaryTree;

import Lib.Tree.Node;
import Lib.Tree.NodeType;

public class BinaryTree {
    public static void main(String[] args) {
        Lib.BinaryTree.BinaryTree<Integer> binaryTree = new Lib.BinaryTree.BinaryTree<>();
        binaryTree.root = new Node<>(3, NodeType.ROOT);
        Node<Integer> n1 = new Node<>(1, NodeType.LEFT_CHILD);
        binaryTree.root.leftChild = n1;
        Node<Integer> n2 = new Node<>(2, NodeType.RIGHT_CHILD);
        binaryTree.root.rightChild = n2;
        n1.leftChild = new Node<>(10, NodeType.LEFT_CHILD);
        n1.rightChild = new Node<>(18, NodeType.RIGHT_CHILD);
        n2.leftChild = new Node<>(11, NodeType.LEFT_CHILD);
        n2.rightChild = new Node<>(8, NodeType.RIGHT_CHILD);

        binaryTree.inOrder(n -> System.out.println(n.data));

        System.out.println("Tree height:\t" + binaryTree.height());
    }
}
